Create restore-backups.md topic

Andriciuc opened the pull request:
Add a new topic:

- How can I store an old key securely? Use Vault as an example here.
- ~~Explain how you can use this “old” principal key to unencrypt your backups.~~ 

Removed files like FAQ and others that do not belong to this PR.

Andriciuc committed:
initial commit that fixes a toc issue, next is the decrypt topic creation

fixed an issue that has been bugging me in tocs first.

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-decrypt-old-backups

Andriciuc committed:
created decrypt backup topic

added basic information and updated ToC, it is under TDE Operations, and updated FAQ response to include link to new topic

Andriciuc committed:
Merge branch 'docs-decrypt-old-backups' of https://github.com/percona/postgres into docs-decrypt-old-backups

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-decrypt-old-backups

dutow commented on contrib/pg_tde/documentation/docs/how-to/decrypt-backups.md:
```diff
@@ -0,0 +1,38 @@
+
+1. Enable the secrets engine by following the [Vault KV documentation](https://developer.hashicorp.com/vault/docs/commands/kv).
+2. Store your current principal key.
+3. When rotating, generate and store a new principal key under a new version or path.
```
This is the only option, you can't overwrite existing keys with pg_tde

dutow commented on contrib/pg_tde/documentation/docs/how-to/decrypt-backups.md:
```diff
@@ -0,0 +1,38 @@
+Here’s a basic workflow using the **KV v2 secrets engine**:
+
+1. Enable the secrets engine by following the [Vault KV documentation](https://developer.hashicorp.com/vault/docs/commands/kv).
+2. Store your current principal key.
```
You don't have to do this directly, just use pg_tde functions normally

dutow commented on contrib/pg_tde/documentation/docs/how-to/decrypt-backups.md:
```diff
@@ -0,0 +1,38 @@
+
+To decrypt a backup made with a previous principal key:
+
+1. Retrieve the correct principal key used at the time of backup from your key management system.
```
No need to this, as long as the key is still there on the same KMS instance. If it's no longer there, then we have the `pg_tde_change_key_provider` utility.

dutow commented on contrib/pg_tde/documentation/docs/how-to/decrypt-backups.md:
```diff
@@ -0,0 +1,38 @@
+1. Retrieve the correct principal key used at the time of backup from your key management system.
+2. Configure `pg_tde` to use this key temporarily.
+3. Restore the backup as usual.
+4. (Optional) After the restore, re-encrypt the data using the latest principal key.
```
No need to reencryption. The entire "decrypt old backups" title is misleading, there isn't anything about decryption here. It's jus about getting old backups running, not removing the encryption from them.

dutow commented on contrib/pg_tde/documentation/docs/how-to/decrypt-backups.md:
```diff
@@ -0,0 +1,38 @@
+To decrypt a backup made with a previous principal key:
+
+1. Retrieve the correct principal key used at the time of backup from your key management system.
+2. Configure `pg_tde` to use this key temporarily.
```
As I said above, if the key is still there as before, the user doesn't have to do anything specific. If the address of the server is changed or something, then we have `pg_tde_change_key_provider`

dutow commented on contrib/pg_tde/documentation/docs/how-to/decrypt-backups.md:
```diff
@@ -0,0 +1,38 @@
+
+Here’s a basic workflow using the **KV v2 secrets engine**:
+
+1. Enable the secrets engine by following the [Vault KV documentation](https://developer.hashicorp.com/vault/docs/commands/kv).
```
The secrets engine is enabled by default

dutow commented on contrib/pg_tde/documentation/docs/how-to/decrypt-backups.md:
```diff
@@ -0,0 +1,38 @@
+- An audit trail for compliance and traceability.
+- Support for versioning and key rotation policies.
+
+## Example: Using HashiCorp Vault to store rotated keys
```
I don't understand this section - this is generic, it would work the same way with kmip...

dutow commented on contrib/pg_tde/documentation/docs/how-to/decrypt-backups.md:
```diff
@@ -0,0 +1,38 @@
+# Handling Key Rotation and Encrypted Backups
+
+Backups created with `pg_tde` are encrypted using the *active principal key* at the time of backup. If this key is rotated and the original version is not retained, those backups **can no longer be decrypted**. This makes **secure key storage** essential for reliable disaster recovery.
```
The entire `active principal key`, and `current principal key` concept is misleading: a single postgres installation can have multiple active/current principal keys at any time, one for each database. It might have just one, but we support multitenancy which means using multiple keys.

dutow commented on contrib/pg_tde/documentation/docs/faq.md:
```diff
@@ -157,7 +157,7 @@ WAL files are encrypted globally across the entire PostgreSQL cluster using the
 Since the encryption happens on the database level, it makes no difference for your tools and applications. They work with the data in the same way.
 
-To restore from an encrypted backup, you must have the same principal encryption key, which was used to encrypt files in your backup.  
+To restore from an encrypted backup, you must have the same principal encryption key, which was used to encrypt files in your backup. For more information, see [Handling Key Rotation and Encrypted Backups](how-to/decrypt-backups.md).
```
> For more information....

In the current state, the new page doesn't add any new information other than this sentence that's already there. It's longer, it has many statements that are not true, but it doesn't have any actual useful information other than what the first sentence here describes.

Andriciuc replied:
Thank you for the thorough review! I agree, in fact I believe we can detail the restoration (not decryption) process a little to make it clearer for newcomers as well. I suggest we keep the topic describing the concept of using a key to unlock an older backup.

Currently we only mention backup restoration in the FAQ without having actually described it anywhere, and while it might be easy to understand I am sure our users will appreciate a bit more information about how it works in the context of pg_tde.

To ensure this I have rewritten the topic and renamed it, and I also added a bit more details to our FAQ such as:

- ## How can I store an old key securely?
- ## What if the address of the key provider changed for one of my old backups?

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-decrypt-old-backups

Andriciuc committed:
renamed and rewrote decrypt to restore

Restoring is more correct in the context of getting an older backup's info. Rewrote the chapter as I added two new question/answers to FAQ with links to our docs, fixed FAQ a bit with better wording.

Andriciuc committed:
Merge branch 'docs-decrypt-old-backups' of https://github.com/percona/postgres into docs-decrypt-old-backups

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-decrypt-old-backups

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-decrypt-old-backups

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-decrypt-old-backups

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into docs-decrypt-old-backups

dutow commented on contrib/pg_tde/documentation/docs/advanced-topics/adv-top-overview.md:
```diff
@@ -4,4 +4,4 @@ This section covers the internal components and tools that power `pg_tde`.
 Use it to understand how encryption is implemented, fine-tune a configuration, leverage advanced CLI tools and functions for diagnostics and customization.
 
-[Architecture](../architecture/index.md){.md-button} [GUC Variables](../variables.md){.md-button} [Functions](../functions.md){.md-button}
+[Architecture](../architecture/architecture.md){.md-button} [GUC Variables](../variables.md){.md-button} [Functions](../functions.md){.md-button}
```
Unrelated to PR title.

dutow commented on contrib/pg_tde/documentation/docs/faq.md:
```diff
@@ -133,17 +133,17 @@ Since the `SET ACCESS METHOD` command drops hint bits and this may affect the pe
 
 * after you enabled the `pg_tde` extension
-* to turn on / off the WAL encryption
+* to turn WAL encryption on or off
```
Not related to pr title (the entire file)

dutow commented on contrib/pg_tde/documentation/docs/how-to/restore-backups.md:
```diff
@@ -0,0 +1,16 @@
+
+To restore an encrypted backup created with an earlier key, ensure the key is still available in your Key Management System (KMS). The backup remains **encrypted** on disk, `pg_tde` uses the correct key to transparently access the data.
+
+Each encrypted table stores metadata that identifies which encryption key was used. When the encrypted data is read, `pg_tde` retrieves the correct key from the configured KMS.
```
Not true: table (internal) keys are stored in the backup. Only the principal key/principal keys get retrieved from the  KMS, which are used to decrypt the internal keys.

Andriciuc replied:
Reworded, removed the metadata mention, just left the Table internal key mention:

The table (internal) keys are stored within the backup. At runtime, `pg_tde` retrieves the principal key(s) from the configured KMS and uses them to decrypt the internal keys, enabling access to the encrypted table data.

dutow commented on contrib/pg_tde/documentation/docs/how-to/restore-backups.md:
```diff
@@ -0,0 +1,16 @@
+
+## How `pg_tde` uses old keys to load backups
+
+* **KMS**: stores the key that was active when the backup was made.
```
possibly multiple *principal* keys

Andriciuc replied:
Updated to: ...stores the principal key(s) that were active...

dutow commented on contrib/pg_tde/documentation/docs/how-to/restore-backups.md:
```diff
@@ -0,0 +1,16 @@
+* **Encrypted backup**: contains data encrypted with the key available at that time.
+* **Current server**: `pg_tde` automatically loads the backup if the key is accessible through the KMS.
+
+At runtime, `pg_tde` reads the key ID from the table metadata and fetches the corresponding key from the KMS.
```
Not true, see above.

Andriciuc replied:
I moved the above corrected paragraph and put it here, feels more natural.

dutow commented on contrib/pg_tde/documentation/docs/how-to/restore-backups.md:
```diff
@@ -0,0 +1,16 @@
+At runtime, `pg_tde` reads the key ID from the table metadata and fetches the corresponding key from the KMS.
+
+!!! note
+    If the key was deleted or the KMS configuration changed, the backup **cannot** be read. Use [pg_tde_change_key_provider](../command-line-tools/pg-tde-change-key-provider.md) to update KMS credentials or endpoints if needed.
```
Note: this only works if the KMS configuration changed. If the key was deleted from everywhere, it can't recover it.

Also, we can have multiple keys, so I would use "if a key"

Andriciuc replied:
I updated to ensure the user understands that key_provider only works if KMS config changed, also added increased importance that if you delete/don't have the required key(s), you can't read the backup

!!! note
    If a required key is missing or was deleted, the backup **cannot** be read.

    If the key still exists but the KMS configuration has changed, you can use [`pg_tde_change_key_provider`](../command-line-tools/pg-tde-change-key-provider.md) to update the configuration.

    However, if the key was deleted from all accessible KMS sources, the encrypted backup is **unrecoverable**.

dutow commented on contrib/pg_tde/documentation/docs/wal-encryption.md:
```diff
@@ -88,4 +88,4 @@ Before turning WAL encryption on, you must follow the steps below to create your
 
 For more technical references related to architecture, variables or functions, see:
-[Technical Reference](advanced-topics/index.md){.md-button}
+[Technical Reference](advanced-topics/adv-top-overview.md){.md-button}
```
unrelated to the pr title

Andriciuc committed:
Delete contrib/pg_tde/documentation/docs/advanced-topics/adv-top-overview.md

removed, not needed in PR

Andriciuc committed:
Delete contrib/pg_tde/documentation/docs/architecture/architecture.md

not required in PR

Andriciuc committed:
Revert "Delete contrib/pg_tde/documentation/docs/architecture/architecture.md"

This reverts commit a9364d2cba66a8aa1fa9de2cc49f25f691634529.

Andriciuc committed:
Revert "Delete contrib/pg_tde/documentation/docs/advanced-topics/adv-top-overview.md"

This reverts commit f6e48d47479bb960baf5361c4ba9ac90314eecbb.

Andriciuc committed:
reverted names for arch and adv topics

not needed to this PR

Andriciuc committed:
reverted wal encrypt and mkdocs names not needed in PR

Andriciuc committed:
Update faq.md

removed the FAQ as it was not PR related

Andriciuc committed:
Update faq.md

small updates to `pg_tde` names in faq

Andriciuc committed:
Update faq.md

again used `` to update pg_tde

Andriciuc committed:
Update restore-backups.md

updated last note to underline that if a key is missing the backup can't be done and moved the second intro paragraph to the old backups ## topic.

Andriciuc committed:
Merge branch 'docs-decrypt-old-backups' of https://github.com/percona/postgres into docs-decrypt-old-backups

Andriciuc committed:
Merge branch 'release-17.5.2' into docs-decrypt-old-backups

Andriciuc committed:
Merge branch 'release-17.5.2' into docs-decrypt-old-backups
